crates/ra_assists/src/add_new.rs

```diff
@@ -0,0 +1,353 @@
+use crate::{Assist, AssistCtx, AssistId, TextRange, TextUnit};
+use hir::{db::HirDatabase, HasSource};
+use ra_db::FilePosition;
+use ra_syntax::{
+    ast::{self, AstNode, NameOwner, TypeAscriptionOwner},
+    T,
+};
+
+pub(crate) fn add_new(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let impl_node = ctx.node_at_offset::<ast::ImplBlock>()?;
```

Comment:
I think making the assist applicable for the whole `impl` block is to much. That is, in a lot of cases the assist be just a false positive.

I think that we should trigger the assist when the cursor is on the struct definition, and that we should either find an existing impl block for it, or create a new one in the same module.

Reply:
I'm finding it surprisingly hard to navigate the API of this library. How do I find all of the impls in the current module (or crate) here?

Reply:
Yeah, the hir API is not yet intuitive and well documented, as we are in the process of figuring stuff out. See `impls_for_def` function for an example of how to search type's impls.